src/cpu/pred/btb_entry.hh

```diff
@@ -0,0 +1,207 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Declaration of a BTB entry and BTB indexing policy.
+ */
+
+#ifndef __CPU_PRED_BTB_ENTRY_HH__
+#define __CPU_PRED_BTB_ENTRY_HH__
+
+#include <vector>
+
+#include "arch/generic/pcstate.hh"
+#include "base/cache/cache_entry.hh"
+#include "base/intmath.hh"
+#include "base/types.hh"
+#include "cpu/static_inst.hh"
+#include "mem/cache/tags/indexing_policies/set_associative.hh"
+#include "params/BTBIndexingPolicy.hh"
+
+namespace gem5::branch_prediction
+{
+
+class BTBTagTypes
+{
+  public:
+    struct KeyType
+    {
+        Addr address;
+        ThreadID tid;
+    };
+    using Params = BaseIndexingPolicyParams;
+};
+
+class BTBEntry : public CacheEntry
+{
+  public:
+    using IndexingPolicy = BTBIndexingPolicy;
+    using KeyType = BTBTagTypes::KeyType;
+    using TagExtractor = std::function<Addr(Addr)>;
+
+    /** Default constructor */
+    BTBEntry(TagExtractor ext)
+        : CacheEntry(ext), inst(nullptr), tid(0)
+    {}
+
+    /** Update the entry in the BTB */
+    void
+    update(ThreadID _tid,
+           const PCStateBase &_target,
+           StaticInstPtr _inst)
+    {
+        tid = _tid;
+        set(target, _target);
+        inst = _inst;
+    }
+
+    /** Match the tag of the BTB entry */
+    bool
+    match(const KeyType &key)
+    {
+        return (match(key.address) && (tid == key.tid));
+    }
+
+    /** Match the tag of the BTB entry */
+    void
+    insert(const KeyType &key)
+    {
+        insert(key.address);
+        tid = key.tid;
+    }
+
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other) : CacheEntry(other)
+    {
+        tid   = other.tid;
+        inst  = other.inst;
+        set(target, other.target);
+    }
+
+    /** The entry's target. */
+    std::unique_ptr<PCStateBase> target;
+
+    /** Pointer to the static branch inst at this address */
+    StaticInstPtr inst;
+
+    /** The entry's thread id. */
+    ThreadID tid;
+
+  private:
+
+    /**
+     * Mark the following as private so as not to call
+     * them directly without tid
+     */
+    using CacheEntry::match;
+    using CacheEntry::insert;
+};
+
+
+class BTBIndexingPolicy : public SetAssociative
```

Comment:
Similarly to the BTBEntry discussion above. We need to pay the price and not inherit from SetAssociative. SetAssociative and the parent BaseIndexingPolicy are assuming an Addr key, whereas here our key is the Addr + Tid pair. I suggest you to have a look at what I did with TaggedIndexingPolicy... This is the hard bit I recon, as it requires you to define a templated SimObject. Happy to provide further assistance if needed

Reply:
@giactra I tried your suggestion and created a BTBIndexingPolicy/BTBSetAssociative policy derived from SimObject. However, I seem to be running into a linking error. Can you please help?

Reply:
Ah, looks like I missed implementing the regenerateAddr method. It works now.